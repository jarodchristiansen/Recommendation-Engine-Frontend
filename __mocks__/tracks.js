export const mockTracks = [
  {
    id: "1",
    name: "Song One",
    subtext: "Artist One",
    image: "/image1.jpg",
    track_id: "1",
    artist_name: "Artist One",
    track_name: "Song One",
    feature_difference: {
      popularity: 1,
      danceability: 1,
      energy: 1,
      valence: 1,
      loudness: 1,
      key: 1,
      speechiness: 1,
    },
    danceability: 1,
    energy: 1,
    key: 1,
    loudness: 1,
    speechiness: 1,
    valence: 1,
    popularity: 1,
    similarity_score: 1,
  },
  {
    id: "2",
    name: "Song Two",
    subtext: "Artist Two",
    image: "/image2.jpg",
    track_id: "2",
    artist_name: "Artist Two",
    track_name: "Song Two",
    feature_difference: {
      popularity: 2,
      danceability: 2,
      energy: 2,
      valence: 2,
      loudness: 2,
      key: 2,
      speechiness: 2,
    },
    danceability: 2,
    energy: 2,
    key: 2,
    loudness: 2,
    speechiness: 2,
    valence: 2,
    popularity: 2,
    similarity_score: 2,
  },
  {
    id: "3",
    name: "Song Three",
    subtext: "Artist Three",
    image: "/image3.jpg",
    track_id: "3",
    artist_name: "Artist Three",
    track_name: "Song Three",
    feature_difference: {
      popularity: 3,
      danceability: 3,
      energy: 3,
      valence: 3,
      loudness: 3,
      key: 3,
      speechiness: 3,
    },
    danceability: 3,
    energy: 3,
    key: 3,
    loudness: 3,
    speechiness: 3,
    valence: 3,
    popularity: 3,
    similarity_score: 3,
  },
];
