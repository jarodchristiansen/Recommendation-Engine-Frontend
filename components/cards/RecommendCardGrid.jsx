import Image from "next/image";
import { PolarAngleAxis, PolarGrid, PolarRadiusAxis, Radar, RadarChart, ResponsiveContainer } from "recharts";

// Generic Card Component for displaying tracks, artists, etc.
const RecommendCardGrid = ({ items, handleItemClick, selectedSongs, type }) => {
  // Checks if an item is already selected
  const isSelected = (track) => {
    return selectedSongs?.some((selectedSong) => selectedSong?.id === track.id);
  };

  // Function to create data for the radar chart from feature differences
  const createRadarData = (item) => {
    console.log({ item });

    const featureDifference = item.feature_difference || {};
    return Object?.keys(featureDifference).map((feature) => {
      const value = Math.abs(featureDifference[feature]);

      return {
        feature,
        value,
      };
    });
  };

//   //   // Function to create data for the bar chart comparison
//   const createBarData = (item, targetFeatures) => {
//     return Object?.keys(targetFeatures).map((feature) => ({
//       feature,
//       target: targetFeatures[feature],
//       recommendation: item[feature],
//     }));
//   };

  return (
    <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6">
      {items.map((item) => (
        <div
          key={item.id}
          onClick={() => handleItemClick(item)}
          className={`group p-4 border rounded-lg cursor-pointer transition-transform transform ${
            isSelected(item) ? "border-blue-500 scale-105" : "border-gray-200"
          } hover:border-blue-500 hover:scale-105`}
        >
          <div className="relative w-full h-48 mb-4">
            <Image
              src={item.image}
              layout="fill"
              objectFit="cover"
              className="rounded-lg"
              alt={`${item.name} ${type}`}
              unoptimized={true}
            />
          </div>
          <h4 className="text-lg font-bold text-gray-900">{item.name}</h4>
          <p className="text-sm text-gray-600">{item.subtext}</p>

          {/* Radar Chart for Feature Differences */}
          <div className="mt-4">
            <h5 className="text-sm font-semibold">Feature Difference</h5>
            <ResponsiveContainer width="100%" height={200}>
              <RadarChart data={createRadarData(item)} outerRadius="80%">
                <PolarGrid />
                <PolarAngleAxis dataKey="feature" />
                <PolarRadiusAxis />
                <Radar
                  name="Difference"
                  dataKey="value"
                  stroke="#f04949"
                  fill="#f04949"
                  fillOpacity={0.6}
                />
              </RadarChart>
            </ResponsiveContainer>
          </div>

          {/* Bar Chart for Feature Comparison */}
          {/* <div className="mt-4">
            <h5 className="text-sm font-semibold">Feature Comparison</h5>
            <ResponsiveContainer width="100%" height={200}>
              <BarChart data={createBarData(item, targetFeatures)}>
                <XAxis dataKey="feature" />
                <YAxis />
                <Tooltip />
                <Bar dataKey="target" fill="#82ca9d" name="Target" />
                <Bar
                  dataKey="recommendation"
                  fill="#8884d8"
                  name="Recommended"
                />
              </BarChart>
            </ResponsiveContainer>
          </div> */}
        </div>
      ))}
    </div>
  );
};

export default RecommendCardGrid;
