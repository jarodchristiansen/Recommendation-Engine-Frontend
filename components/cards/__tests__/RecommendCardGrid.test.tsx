import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import RecommendCardGrid from "../RecommendCardGrid";

import { mockTracks } from "../../mocks/tracks";

// import "@testing-library/jest-dom/extend-expect"; // for extended matchers

describe("RecommendCardGrid", () => {
  const handleItemClick = jest.fn(); // Mock function for handling clicks
  const selectedSongs = [mockTracks[0]]; // Pre-select the first song

  beforeAll(() => {
    global.ResizeObserver = class {
      observe() {}
      unobserve() {}
      disconnect() {}
    };
  });

  it("renders the items correctly", () => {
    render(
      <RecommendCardGrid
        items={mockTracks}
        handleItemClick={handleItemClick}
        selectedSongs={[]}
        type="track"
      />
    );

    // Check if the items are rendered
    expect(screen.getByText("Song One")).toBeInTheDocument();
    expect(screen.getByText("Song Two")).toBeInTheDocument();
    expect(screen.getByText("Song Three")).toBeInTheDocument();
  });

  it("applies selected class when a song is selected", () => {
    render(
      <RecommendCardGrid
        items={mockTracks}
        handleItemClick={handleItemClick}
        selectedSongs={selectedSongs}
        type="track"
      />
    );

    // Check if the selected item has the correct class applied
    const selectedCard = screen.getByText("Song One").closest("div");
    expect(selectedCard).toHaveClass("border-blue-500 scale-105");
  });

  it("calls handleItemClick when an item is clicked", () => {
    render(
      <RecommendCardGrid
        items={mockTracks}
        handleItemClick={handleItemClick}
        selectedSongs={[]}
        type="track"
      />
    );

    // Simulate clicking on the second song
    const songTwo = screen.getByText("Song Two").closest("div");
    fireEvent.click(songTwo!);

    // Check if the click handler was called with the right item
    expect(handleItemClick).toHaveBeenCalledWith(mockTracks[1]);
  });

  it("renders correctly with no items", () => {
    render(
      <RecommendCardGrid
        items={[]}
        handleItemClick={handleItemClick}
        selectedSongs={[]}
        type="track"
      />
    );

    // Ensure no items are rendered
    expect(screen.queryByText("Song One")).not.toBeInTheDocument();
    expect(screen.queryByText("Song Two")).not.toBeInTheDocument();
    expect(screen.queryByText("Song Three")).not.toBeInTheDocument();
  });
});
