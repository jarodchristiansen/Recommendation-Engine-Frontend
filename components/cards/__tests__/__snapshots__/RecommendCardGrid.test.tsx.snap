// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendCardGrid matches the snapshot of the RecommendCardGrid 1`] = `
<div>
  <div
    class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6"
  >
    <div
      class="group p-4 border rounded-lg cursor-pointer transition-transform transform border-gray-200 hover:border-blue-500 hover:scale-105"
    >
      <div
        class="relative w-full h-48 mb-4"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Song One track"
            class="rounded-lg"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
          />
          <noscript />
        </span>
      </div>
      <h4
        class="text-lg font-bold text-gray-900"
      >
        Song One
      </h4>
      <p
        class="text-sm text-gray-600"
      >
        Artist One
      </p>
      <div
        class="mt-4"
      >
        <h5
          class="text-sm font-semibold"
        >
          Feature Difference
        </h5>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 200px; min-width: 0;"
        />
      </div>
    </div>
    <div
      class="group p-4 border rounded-lg cursor-pointer transition-transform transform border-gray-200 hover:border-blue-500 hover:scale-105"
    >
      <div
        class="relative w-full h-48 mb-4"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Song Two track"
            class="rounded-lg"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
          />
          <noscript />
        </span>
      </div>
      <h4
        class="text-lg font-bold text-gray-900"
      >
        Song Two
      </h4>
      <p
        class="text-sm text-gray-600"
      >
        Artist Two
      </p>
      <div
        class="mt-4"
      >
        <h5
          class="text-sm font-semibold"
        >
          Feature Difference
        </h5>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 200px; min-width: 0;"
        />
      </div>
    </div>
    <div
      class="group p-4 border rounded-lg cursor-pointer transition-transform transform border-gray-200 hover:border-blue-500 hover:scale-105"
    >
      <div
        class="relative w-full h-48 mb-4"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Song Three track"
            class="rounded-lg"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
          />
          <noscript />
        </span>
      </div>
      <h4
        class="text-lg font-bold text-gray-900"
      >
        Song Three
      </h4>
      <p
        class="text-sm text-gray-600"
      >
        Artist Three
      </p>
      <div
        class="mt-4"
      >
        <h5
          class="text-sm font-semibold"
        >
          Feature Difference
        </h5>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 200px; min-width: 0;"
        />
      </div>
    </div>
  </div>
</div>
`;
